File: SecondYear/Akash-Tiwari22_AkashTiwari_2226cs1058_2/Week_2/generateParenthesis.cpp
void generateParenthesis(int n, int openCount, int closeCount, string current, vector<string>& result) {
    if (current.length() == 2 * n) {
        result.push_back(current);
        return;
    }

    if (openCount < n) {
        generateParenthesis(n, openCount + 1, closeCount, current + '(', result);
    }

    if (closeCount < openCount) {
        generateParenthesis(n, openCount, closeCount + 1, current + ')', result);
    }
}
vector<string> validParenthesis(int n){
    // Write your code here.
     vector<string> result;
    generateParenthesis(n, 0, 0, "", result);
    return result;

}

File: SecondYear/aayushraghav93_AayushRaghav_2226CSE(AI)1108_2/Week_2/GenerateParenthesis.java
import java.util.* ;
import java.io.*; 
public class GenerateParenthesis {
    public static ArrayList<String> validParenthesis(int n){
        // Write your code here.
        ArrayList<String> result = new ArrayList<>();
        generateParenthesisHelper(n, n, "", result);
        return result;
    }

    private static void generateParenthesisHelper(int left, int right, String current, List<String> result) {
        if (left == 0 && right == 0) {
            result.add(current);
            return;
        }

        if (left > 0) {
            generateParenthesisHelper(left - 1, right, current + "(", result);
        }

        if (right > left) {
            generateParenthesisHelper(left, right - 1, current + ")", result);
        }
    } 
}